apps/jupyter/extensions/notebook-runner/src/commands/grade.ts

```diff
@@ -0,0 +1,156 @@
+import { JupyterFrontEnd } from "@jupyterlab/application";
+import { INotebookTracker, NotebookPanel } from "@jupyterlab/notebook";
+import { Contents, ContentsManager } from "@jupyterlab/services";
+import { IDocumentManager } from "@jupyterlab/docmanager";
+import { NotebookRunnerState } from "../notebook-runner-state";
+import { executeRunNotebookCommand, runGradingCommand } from "../command";
+import { EmbeddedPythonCallbacks } from "../iframe-api";
+import { OtterGradingResults } from "../grading-results";
+import {
+  writeBinaryToVirtualFilesystem,
+  writeJsonToVirtualFilesystem,
+} from "../utils";
+
+const binaryResultsPath = "/results.pkl";
+
+const createOnNewNotebookListener =
+  (app: JupyterFrontEnd, state: NotebookRunnerState) =>
+  async (
+    _notebookTracker: INotebookTracker,
+    notebookPanel: NotebookPanel,
+  ): Promise<void> => {
+    // close all other notebooks
+    if (!state.allowNextNotebookInParallel) {
+      const widgets = app.shell.widgets("main");
+
+      for (const widget of widgets) {
+        if (widget instanceof NotebookPanel && widget !== notebookPanel) {
+          console.log("Closing notebook:", widget);
+          await widget.context.save();
+          widget.close();
+        }
+      }
+    }
+    state.allowNextNotebookInParallel = false;
+  };
+
+export const registerGradeCommand = (
+  state: NotebookRunnerState,
+  app: JupyterFrontEnd,
+  notebookTracker: INotebookTracker,
+  contentsManager: ContentsManager,
+  documentManager: IDocumentManager,
+): void => {
+  app.commands.addCommand(runGradingCommand, {
+    label: "Run Grading",
+    execute: async () => {
+      console.debug(`Waiting for otter session kernel to be available`);
+      const kernel = await state.getOtterKernel();
+
+      console.debug(`Transferring autograder to virtual filesystem`);
```

Comment:
The virtual filesystem is an implementation detail. The important thing here is that we're loading the autograder for grading.